Documentation for memory optimization and micro iterations topics (#1264)
AntaresSimulatorTeam/antares-xpansion: docs/reference/benders/micro-iterations.md

Reviewer: objective $c$ constraints $a_1$, etc

faster to solve.

This relaxation is solved a first time, and the resulting solution is handed to
a verification mechanism whose role is to check, outside of the optimization
itself, whether the constraints that were left out of the model are satisfied
by the current solution. The constraints found to be violated are injected into
the subproblem, which is then re-solved with this enriched formulation.

This solve–check–inject sequence constitutes one *micro-iteration*, and it is
repeated until the verification step detects no further violation. At that
point, the solution of the relaxed problem is feasible for the complete
formulation — and therefore optimal for it, since removing constraints can only
relax the problem.

The subproblem thus returns exactly the same optimal value and dual information
as if it had been solved with all constraints from the start, but the solver
only ever handles the constraints that actually matter: in practice, a large
share of the operational constraints are inactive at the optimum, and the
successive relaxations remain far smaller than the full problem.

![Micro Iterations](../../assets/media/micro_iterations.png)

> **`Warm start`**

Reviewer: ```suggestion
## Warm start
```

Because Benders decomposition solves the subproblems repeatedly across
successive iterations, the micro-iterations scheme can carry its learning over
from one iteration to the next. Rather than rebuilding each subproblem from its
minimal relaxed formulation and re-discovering the relevant constraints from
scratch, the warm-start mechanism initializes the subproblem with the
constraints that had been injected during the previous iteration's
solve–check–inject sequence. These constraints are known to have been active
(i.e. binding or violated) for the previous master solution, and since
successive master solutions usually remain close to one another, a large share
of them are still needed. Starting from this enriched formulation typically
requires far fewer micro-iterations to reach feasibility, at the cost of a
slightly larger initial problem. The trade-off is favorable in practice: the
constraints carried over from the previous iteration are precisely those most
likely to matter again, so the relaxation remains small while skipping most of

> the redundant solve–check–inject cycles.

Reviewer: Be more nuanced, this may not be the case every time in practice, say that the tool has an option to activate that, but it is the user responsibility to assess whether warm starting micro iterations is faster than resolving them from scratch (as for some use cases, the active constraints may be very different from one benders iteration to another, conversely to what is dtated here)

violation. Compared to a standard study, the following elements must be added:

- **`MICRO_ITERATIONS: true`**  in the `options.json` file, see [Settings of benders executable](./options.md) — the simulation variable that activates the
  micro-iterations mode. Without it, the study is solved with the classical
  scheme in which each subproblem carries its full set of constraints from the
  start.

- **MPS/SVF files for the removed constraints** — in addition to the MPS/SVF
  files representing the initial (relaxed) subproblem optimization problems,
  the study must provide MPS/SVF files describing the constraints that were
  removed from the initial formulation. These files are the pool from which
  violated constraints are injected into the subproblems at each
  micro-iteration.

    To make the role of these files concrete, consider a subproblem with
    objective $O$ and three constraints $C_1$, $C_2$ and $C_3$. With the
    classical scheme, the subproblem is written as a single optimization
    problem:

    $$
    \begin{aligned}
        \min_{x} \quad & O(x) \\
        \text{s.t.} \quad & {\scriptstyle C_1(x), \; C_2(x), \; C_3(x)}
    \end{aligned}

>     $$

Reviewer: We only manipulate linear problem, so the O(x) , C(x) are too general, and the classical notation use `c` for cost and `a` for constraints
```suggestion
    $$
    \begin{aligned}
        \min_{x} \quad & c^\top x \\
        \text{s.t.} \quad &  a_1^\top x =b_1\\
        & a_2^\top x = b_2\\
        & a_3^\top x = b_3
    \end{aligned}
    $$
```

    In micro-iterations mode, this problem is split into the initial relaxed
    subproblem — which only carries $C_1$ — and the constraints removed from

>     it, $C_2$ and $C_3$, described in separate MPS/SVF files:

Reviewer: ```suggestion
    subproblem — which only carries $a_1$ — and the constraints removed from
    it, $a_2$ and $a_3$, described in separate MPS/SVF files:
```

    $$
    \underbrace{
    \begin{aligned}
        \min_{x} \quad & O(x) \\
        \text{s.t.} \quad & {\scriptstyle C_1(x)}
    \end{aligned}
    }_{\text{initial minimal subproblem}}
    \qquad\qquad\Big|\qquad\qquad
    \underbrace{
    \begin{aligned}
        & {\scriptstyle C_2(x), \; C_3(x)}
    \end{aligned}
    }_{\text{constraints file with null objective function}}

>     $$

Reviewer: Check rendering, I did the suggestion by hand without testing
```suggestion
    $$
    \underbrace{
    \begin{aligned}
        \min_{x} \quad & c^\top x \\
        \text{s.t.} \quad & a_1^\top x = b_1
    \end{aligned}
    }_{\text{initial minimal subproblem}}
    \qquad\qquad\Big|\qquad\qquad
    \underbrace{
    \begin{aligned}
        & a_2^\top x =b_2 \\
        & a_3^\top x = b_3
    \end{aligned}
    }_{\text{constraints file with null objective function}}
    $$
```

successive relaxations remain far smaller than the full problem.

![Micro Iterations](../../assets/media/micro_iterations.png)

## Launching a study with micro-iterations

The micro-iterations mode is enabled through the simulation settings by setting
the variable `MICRO_ITERATIONS: true`. When this mode is active, the study must
be enriched with additional inputs that describe the constraints excluded from
the initial subproblems and provide the machinery needed to detect their
violation. Compared to a standard study, the following elements must be added:

- **`MICRO_ITERATIONS: true`** — the simulation variable that activates the
  micro-iterations mode. Without it, the study is solved with the classical
  scheme in which each subproblem carries its full set of constraints from the
  start.

- **MPS/SVF files for the removed constraints** — in addition to the MPS/SVF
  files representing the initial (relaxed) subproblem optimization problems,
  the study must provide MPS/SVF files describing the constraints that were
  removed from the initial formulation. These files are the pool from which
  violated constraints are injected into the subproblems at each
  micro-iteration.

> - **A `plugin_inputs` folder** — this folder contains the artefacts needed to

Author: after discussion, we concluded that we don't need to do this since everything in this folder is plugin related
